### DISCUSSION TITLE
Is cluster-wide "watch" on Secrets truly required?

### DISCUSSION THREAD

**Discussion by schahal** [istio/istio#53609] [Q&A]
## Context

In a multi-primary, multi-cluster setup, we see the following in our Istiod Pods (e.g., on `cluster-a` istiod pods):

```
2024-10-22T12:21:35	error	watch error in cluster cluster-b: unknown (get secrets)
2024-10-22T12:21:47	error	watch error in cluster cluster-c: unknown (get secrets)
```

This is mitigated by adding `watch` to a ClusterRole (rather than a namespaced Role) for the K8s RBAC (e.g., similar to the default reader role).

We'd like to avoid this for security concerns over what may be able to bind to that role.

## The Question

Is there room for enhancements to make Secrets watchers just be on the root/installation namespace? 

I believe the source code offending code of the above is https://github.com/istio/istio/blob/4dc8332b1b097292efa16db23df25ca9fdc6a48f/pilot/pkg/credentials/kube/secrets.go#L79

## Similar Discussions/Issues

(I can't tell if there's a definitive answer to the question in any of these - my guess is _**yes**_ for e.g., Gateway APIs to work?)

- https://github.com/istio/istio/discussions/48587 (will [the PR](https://github.com/istio/istio/pull/53175) from that discussion fix this?)
-  https://github.com/istio/istio/issues/46701 (using `ISTIO_WATCH_NAMESPACE` didn't get me what I wanted, that's more for multi-tenanted setups)

**Comment by howardjohn** [comment]
Is it required for _all_ users who don't use every feature - no. Is it required for full functionality of core functionality -- yes, gateways required it (all types, not just Gateway API).

**Reply by schahal** [reply]
Thanks @howardjohn ! So we can technically get away by not adding the `watch` to the ClusterRole (EDIT: for the roles used for multi-cluster auth).

But, the downside there is that we will be littered with harmless "errors" in our logs.

Is there a workaround where Istio won't spew those errors in that case (that you're aware of)?

**Reply by diranged** [reply]
@howardjohn I'd love to know more about why and what secrets Istio is actually looking at for cross-cluster communication?
